### PR TITLE
14670 Added ledger component for restoration extension filings

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#14670

*Description of changes:*

First commit:
- app version = 5.11.6
- imported latest shared corp-type-module
- imported latest Vuetify
- replaced some getLegalName with getEntityName
- added some getLegalName fallback strings
- fixed some typing errors
- added handling for Restoration Extension filings
- added Limited Restoration Extension Filing component
- renamed isAdminFreeze -> isAdminFrozen
- changed some Enum Mixin calls to Enum Utilities
- deleted obsolete Enum Mixin methods
- cleaned up EntityInfo first and second lines
- moved EntityInfo tombstone data to new component
- renamed some IDs
- added some getters to get rid of `@State` usage
- deleted some unneeded getter refs
- created Tombstone component
- added getEntityName getter
- moved some computed to store getters
- created/updated/fixed unit tests
- misc cleanup

Second commit:
- moved Entity Info header to new component
- moved Entity Info menu to new component
- renamed Tombstone -> EntityDefinitions
- misc cleanup
- fixed a getter name
- updated/created unit test suites for affected components

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: src/views/Dashboard.vue [hunk 1]
```diff
@@ -114,50 +117,52 @@
               </v-card>
             </section>
 
-            <!-- Current Directors-->
-            <section v-if="!isSoleProp && !isPartnership">
+            <!-- Proprietor / Partners -->
+            <section v-if="isSoleProp || isPartnership">
```
Comment (severinbeauvais) [inline src/views/Dashboard.vue:121]: I just moved the "if True" code above the "else" code.

File: src/components/EntityInfo.vue [hunk 2]
```diff
@@ -331,27 +254,22 @@ export default class EntityInfo extends Vue {
     this.emitConfirmDissolution()
   }
 
-  /** Navigates to the Auth UI to update business profile. */
-  protected editBusinessProfile (): void {
-    navigate(this.getBusinessProfileUrl)
-  }
```
Comment (severinbeauvais) [inline src/components/EntityInfo.vue:337]: This is moved to Tombstone component.

File: src/components/EntityInfo.vue [hunk 3]
```diff
@@ -406,32 +318,6 @@ menu > span + span {
...
-  display: none;
-}
-
-#change-email-button,
-#change-phone-button {
-  height: 1.125rem;
-  padding: 0.25rem 0.5rem;
-  margin-top: -0.125rem;
-  margin-left: 0.125rem;
-}
```
Comment (severinbeauvais) [inline src/components/EntityInfo.vue:433]: This is moved to Tombstone component.

File: src/components/Dashboard/FilingHistoryList.vue [hunk 4]
```diff
@@ -686,6 +710,12 @@ export default class FilingHistoryList extends Vue {
         item.expiry = filing.data.restoration?.expiry
       }
 
+      // add properties for limited restoration extensions
+      if (EnumUtilities.isTypeRestorationExtension(filing)) {
+        item.isTypeLimitedRestorationExtension = true
+        item.expiry = filing.data.restoration?.expiry
+      }
+
```
Comment (jonathan-longe) [inline src/components/Dashboard/FilingHistoryList.vue:718]: Would it be possible to add these attributes inside the `LimitedRestorationExtension` filing history component?

File: src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import { Component, Prop } from 'vue-property-decorator'
+import { ContactInfo } from '@/components/common'
+import { HistoryItemIF } from '@/interfaces'
+import { DateUtilities } from '@/services'
+
+@Component({
+  components: { ContactInfo }
+})
+export default class LimitedRestorationExtensionFiling extends Vue {
+  /** The subject filing. */
```
Comment (jonathan-longe) [inline src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue:28]: Since this is a new component, what do you think about using the more the Options API rather than the Class API so it's natively supported by Vue3.

File: src/components/Dashboard/StaffNotation.vue [hunk 6]
```diff
@@ -76,7 +76,7 @@
             <v-list-item
               data-type="registrars-order"
               @click="showRegistrarsOrderDialog()"
-              :disabled="disabled || isAdminFreeze"
+              :disabled="disabled || isAdminFrozen"
```
Comment (jonathan-longe) [inline src/components/Dashboard/StaffNotation.vue:79]: Is this change related to the ticket?

File: src/components/EntityInfo.vue [hunk 7]
```diff
@@ -4,40 +4,40 @@
...
-                  small label text-color="white"
-                >
-                  LIMITED RESTORATION
-                </v-chip>
-              </span>
+            <template v-if="!!businessId">
+              <!-- First line -->
+              <div id="entity-legal-name" aria-label="Entity Legal Name">
+                {{ getEntityName || 'Unknown Name' }}
+              </div>
```
Comment (jonathan-longe) [inline src/components/EntityInfo.vue:11]: What do you think of creating two components, `EntityInfoBusinessId` and `EntityInfoTempRegId`?  It feels like we're mixing concerns by combining the two.

File: src/components/EntityInfo.vue [hunk 8]
```diff
@@ -141,62 +141,7 @@
...
-              <dd id="entity-business-phone" class="cursor-pointer" @click="editBusinessProfile()">
-                <span>{{fullPhoneNumber || 'Not Available'}}</span>
-                <v-btn id="change-phone-button" small text color="primary">
-                  <v-icon small>mdi-pencil</v-icon>
-                  <span>Change</span>
-                </v-btn>
-              </dd>
-            </template>
-          </dl>
+          <Tombstone />
```
[the remark this answers is not in the document]
Comment (jonathan-longe) [reply src/components/EntityInfo.vue:144]: I think it's recommended to use two words for component names.  What about `EntityKeyFacts`?  See:  https://v2.vuejs.org/v2/style-guide/?redirect=true#Multi-word-component-names-essential

File: src/components/Tombstone.vue [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+<script lang="ts">
+import Vue from 'vue'
+import { Component } from 'vue-property-decorator'
+import { Getter } from 'vuex-class'
+import { navigate } from '@/utils'
+import { DateUtilities } from '@/services'
+
+@Component({})
+export default class Tombstone extends Vue {
```
Comment (jonathan-longe) [inline src/components/Tombstone.vue:67]: What about using the Options API so it's compatible with Vue3?

File: src/components/Tombstone.vue [hunk 10]
```diff
@@ -0,0 +1,171 @@
+<template>
+  <dl id="tombstone">
+    <!-- Registration Date -->
+    <template v-if="!!registrationDate">
+      <dt class="mr-2">Registration Date:</dt>
+      <dd id="registration-date">{{ registrationDate }}</dd>
+    </template>
```
Comment (jonathan-longe) [inline src/components/Tombstone.vue:7]: What do you think about creating a component for the items?  It looks like you have only two variants, a display version and an edit version.   The benefit is that if the business needs to modify the key fact displayed all we need to do is change the data supplied.  We don't have to add an if-then.

File: src/components/Tombstone.vue [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  @Getter getFoundingDate!: Date
+  @Getter getFullPhoneNumber!: string
+  @Getter getIdentifier!: number
+  @Getter getNameRequest!: any
+  @Getter isFirm!: boolean
+
+  /** The Business ID string, if available. */
+  get businessId (): string {
+    return sessionStorage.getItem('BUSINESS_ID')
+  }
```
Comment (jonathan-longe) [inline src/components/Tombstone.vue:80]: What do you think about passing the `businessId` in as a prop?  If we eventually want to have the businessId passed to the view as route parameter, this makes this refactoring easier in the future.

File: src/mixins/enum-mixin.ts [hunk 12]
```diff
@@ -1,32 +1,12 @@
...
-  /** DEPRECATED Returns True if item status is Corrected. */
-  isStatusCorrected (item: any): boolean {
-    return (item.status === FilingStatus.CORRECTED)
-  }
-
-  /** DEPRECATED Returns True if item status is Deleted. */
-  isStatusDeleted (item: any): boolean {
-    return (item.status === FilingStatus.DELETED)
-  }
-
```
Comment (jonathan-longe) [inline src/mixins/enum-mixin.ts:29]: Nice!  Get rid of those deprecated methods.

File: src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue (hunk 5 above)
Comment (severinbeauvais) [reply src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue:28]: A quick Google search suggests Class API is well supported in Vue3:
https://github.com/vuejs/vue-class-component/issues/402
https://levelup.gitconnected.com/from-vue-class-component-to-composition-api-ef3c3dd5fdda
https://medium.com/@takeit_/vue-3-composition-api-with-typescript-class-components-dbe7cc9bb468

In THIS PR, I am following existing code. In my next (refactoring) PR, I will look at this and change it over if it's not too much of a PITA, ok?

File: src/components/Dashboard/StaffNotation.vue (hunk 6 above)
Comment (severinbeauvais) [reply src/components/Dashboard/StaffNotation.vue:79]: No.

`adminFreeze` is the name of the variable, whereas the getter returns the state.

File: src/components/EntityInfo.vue (hunk 7 above)
Comment (severinbeauvais) [reply src/components/EntityInfo.vue:11]: I'll look into this but it is a bit beyond the scope of this PR.

File: src/components/Tombstone.vue (hunk 9 above)
Comment (severinbeauvais) [reply src/components/Tombstone.vue:67]: See other comment. I want to move forward this way for now. This shouldn't be much additional work to change later if we change all other components over at the same time.

File: src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue (hunk 5 above)
Comment (jonathan-longe) [reply src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue:28]: Apologies, that's my mistake, when we discussed this issue previously (see below) I shared a link that I misread to read the Class API support had been abandoned in Vue3.  

image

File: src/components/Tombstone.vue (hunk 9 above)
Comment (jonathan-longe) [reply src/components/Tombstone.vue:67]: You're correct, the Class API is supported in Vue3.  Please ignore the suggestion.

File: src/components/Tombstone.vue (hunk 10 above)
Comment (severinbeauvais) [reply src/components/Tombstone.vue:7]: There is no edit version here; only items that may or may not apply to the currently loaded entity.

Sorry if I don't understand your suggestion. What do you mean?

File: src/components/Tombstone.vue (hunk 11 above)
Comment (severinbeauvais) [reply src/components/Tombstone.vue:80]: Good idea!

File: src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue [hunk 13]
```diff
@@ -0,0 +1,47 @@
+<template>
+  <div v-if="filing" class="limited-restoration-extension-filing body-2">
+    <h4>Extension of Limited Restoration</h4>
+
+    <p>
+      The period of restoration was successfuly extended and is active
+      <strong>until {{ expiryDateFriendly }}</strong>. At the end of the extended limited
```
Comment (JazzarKarim) [inline src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue:7]: Nice! This will help my extension section ticket! 😄

**Review by JazzarKarim — APPROVED** [review]
Lots and LOTS of learning for me. Especially since it's Filings UI. Awesome stuff!

**Review by ketaki-deodhar — APPROVED** [review]
Thanks for the opportunity to review and learn!

File: src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue (hunk 5 above)
Comment (severinbeauvais) [reply src/components/Dashboard/FilingHistoryList/LimitedRestorationExtensionFiling.vue:28]: Are you OK if I don't change this component to Options API, and if so can you please resolve this conversation? TIA

File: src/components/Tombstone.vue (hunk 9 above)
Comment (severinbeauvais) [reply src/components/Tombstone.vue:67]: Are you OK to resolve this conversation?

File: src/components/EntityInfo.vue (hunk 7 above)
Comment (severinbeauvais) [reply src/components/EntityInfo.vue:11]: I'm working on this (and old "Tombstone" sub-component) atm. I think you'll like what I've done 😁

File: src/components/Dashboard/FilingHistoryList.vue (hunk 4 above)
Comment (severinbeauvais) [reply src/components/Dashboard/FilingHistoryList.vue:718]: Are you OK with leaving this until Phase 2 of this work (next PR)?

File: src/components/EntityInfo.vue (hunk 7 above)
Comment (severinbeauvais) [reply src/components/EntityInfo.vue]: Please see new commit and resolve this conversation if you are satisfied.

File: src/components/Tombstone.vue (hunk 11 above)
Comment (severinbeauvais) [reply src/components/Tombstone.vue]: Please see new commit and resolve this conversation if you are satisfied.

File: src/components/Dashboard/StaffNotation.vue (hunk 6 above)
Comment (jonathan-longe) [reply src/components/Dashboard/StaffNotation.vue:79]: Yes, you bet, just clarifying to make sure I wasn't missing something important.

File: src/components/EntityInfo/EntityHeader.vue [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+          <v-chip class="primary mt-n1 ml-3 pointer-events-none font-weight-bold"
+            small label text-color="white">LIMITED RESTORATION</v-chip>
+        </span>
+
+        <span id="authorized-to-continue-out" v-if="isAuthorizedToContinueOut">
+          <v-chip class="primary mt-n1 ml-3 pointer-events-none font-weight-bold"
+            small label text-color="white">AUTHORIZED TO CONTINUE OUT</v-chip>
+        </span>
+      </div>
+    </template>
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityHeader.vue:24]: I know this may seem like a pain, but can we separate the code for entities with a `businessId` from entities with a `tempRegNumber` into their own components?  I eventually would like to have separate dashboard routes for entities with a `businessID` and entities with a `tempRegNumber`.  Separating these concerns will make this easier.

File: src/components/EntityInfo/EntityHeader.vue [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+      <!-- Title -->
+      <div id="ia-reg-name" aria-label="Incorporation Application or Registration Entity Name">
+        {{ getEntityName || 'Unknown Name' }}
+      </div>
+
+      <!-- Subtitle -->
+      <div id="ia-reg-description" aria-label="Incorporation Application or Registration Description">
+        {{ iaRegDescription }}
+      </div>
+    </template>
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityHeader.vue:36]: Here's the template for the EntityHeaderTempReg

File: src/components/EntityInfo/EntityDefinitions.vue [hunk 16]
```diff
@@ -0,0 +1,168 @@
+<template>
+  <dl id="entity-definitions">
+    <!-- Registration Date -->
+    <template v-if="!!registrationDate">
+      <dt class="mr-2">Registration Date:</dt>
+      <dd id="registration-date">{{ registrationDate }}</dd>
+    </template>
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityDefinitions.vue:7]: Apologies for my cryptic comment earlier!  What do you think of making the following a component?

```javascript
  <template>
      <dt class="mr-2">{{ title }}:</dt>
      <dd :id="id">{{ value }}</dd>
    </template>
```

And this a component too?

```javascript
   <template>
      <dt class="mr-2">{{ title }}:</dt>
      <dd :id="id" class="cursor-pointer" @click="editBusinessProfile()">
        <span>{{ value }}</span>
        <v-btn :id="'change' + id + 'button'" small text color="primary">
          <v-icon small>mdi-pencil</v-icon>
          <span>Change</span>
        </v-btn>
      </dd>
    </template>
```

I know this may seem crazy to create tiny components like this but I think you'll see it makes our code more flexible.  By displaying these components by iterating over an array of data, it makes it very easy to change in the future if the business changes the requirements.

File: src/components/EntityInfo/EntityMenu.vue [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+    <span v-if="isHistorical">
+      <v-chip
+        id="historical-chip"
+        class="primary mt-n1 ml-4 pointer-events-none" small label text-color="white"
+      >
+        <strong>HISTORICAL</strong>
+      </v-chip>
+      <span class="font-14 mx-3">{{ getReasonText || 'Unknown Reason' }}</span>
+    </span>
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityMenu.vue:20]: This looks like a separate concern that deserves it's own component.

File: src/components/EntityInfo/EntityMenu.vue [hunk 18]
```diff
@@ -0,0 +1,186 @@
+<template>
+  <menu id="entity-menu">
+    <!-- Staff Comments -->
+    <span v-if="!!businessId && isAllowed(AllowableActions.ADD_STAFF_COMMENT)">
+      <StaffComments
+        :axios="axios"
+        :businessId="businessId"
+        :maxLength="2000"
+      />
+    </span>
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityMenu.vue:10]: Rather than writing conditionals for each allowable action, would it be possible to iterate over the allowable actions returned by the API and then use a `<component :is="">` to render these menu items dynamically?   It seems to me that this would make our code easier to extend and maintain.

File: src/components/EntityInfo/EntityMenu.vue [hunk 19]
```diff
@@ -0,0 +1,186 @@
+<template>
+  <menu id="entity-menu">
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityMenu.vue:2]: What do you think about either moving `<menu>` back to the parent or creating a wrapper with a slot so the subcomponents can be listed like we were thinking about doing with the `YourCompany` component?

File: src/components/EntityInfo/EntityMenu.vue [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+    <!-- Download Business Summary -->
+    <span v-if="isAllowed(AllowableActions.DOWNLOAD_BUSINESS_SUMMARY)">
+      <v-tooltip top content-class="top-tooltip" transition="fade-transition">
+        <template v-slot:activator="{ on }">
+          <v-btn
+            small text color="primary"
+            id="download-summary-button"
+            @click="emitDownloadBusinessSummary()"
+            v-on="on"
+          >
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityMenu.vue:73]: When this button is clicked, it:
- calls a method, `emitDownloadBusinessSummary()`
- which emits an event called, `downloadBusinessSummary`
- which is received by the parent, "EntityInfo"
- which calls another method called, `emitDownloadBusinessSummary`
- which emits another method called `downloadBusinessSummary`
- which is received by the parent, "App.vue"
- which calls another method called, `downloadBusinessSummary`
- which calls the API and downloads the requested document

Would it be possible to call a store action instead or, if you don't have time, create a refactoring ticket that explains the issue and describes a resolution?

File: src/components/EntityInfo/EntityMenu.vue [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+      <span v-if="isAllowed(AllowableActions.DISSOLVE_COMPANY)">
+        <v-tooltip top content-class="top-tooltip" transition="fade-transition">
+          <template v-slot:activator="{ on }">
+            <v-btn
+              small text color="primary"
+              id="dissolution-button"
+              :disabled="hasBlocker"
+              @click="promptDissolve()"
+              v-on="on"
+            >
```
Comment (jonathan-longe) [inline src/components/EntityInfo/EntityMenu.vue:53]: This button and several other buttons in this component appear to also have a long chains of emits that are passed back to the parent and then the grandparent.  Can we either reduce the length of the chains or create tech debt tickets to refactor this?

File: src/components/EntityInfo/EntityHeader.vue (hunk 14 above)
Comment (severinbeauvais) [reply src/components/EntityInfo/EntityHeader.vue:24]: I'd rather not, at this time anyway. Separate routes sound a good idea for the future, but it's still fair to have separate views use common components that check a variable to display A vs B. Also, I like the fact that the Entity Info's header content is all in one file. Also, I don't want to assume that businesses vs pre-businesses (temp reg number) will continue to be handled separately in the current way. Also, this isn't a complex component to break apart in the future if the need arises (the need isn't here yet). Finally, there are many higher-priority, higher-gain things to work first/instead.

Here's an idea of what we'll be up against if we try to split up businesses from pre-businesses -- the A vs B code is in a lot of places so that we reuse components.

image

image

File: src/components/EntityInfo/EntityHeader.vue (hunk 15 above)
Comment (severinbeauvais) [reply src/components/EntityInfo/EntityHeader.vue:36]: If you don't want to see this template when you're in this file, code-fold it.

At the moment, I prefer to keep like content together instead of managing more component files. Also see my previous comment on this file.

File: src/components/EntityInfo/EntityDefinitions.vue (hunk 16 above)
Comment (severinbeauvais) [reply src/components/EntityInfo/EntityDefinitions.vue:7]: We don't need that flexibility yet. Let's change it when/if we need to.

Right now I prefer to keep this entire "description list" together.

File: src/components/EntityInfo/EntityMenu.vue (hunk 18 above)
Comment (severinbeauvais) [reply src/components/EntityInfo/EntityMenu.vue:10]: That's a good idea but I doubt it: the possible actions (and whether they're allowed or not) are sprinkled into different places in the code/components. It boils down to our desire to show a consistent page with various actions but those actions are in different places (header, right-side components, staff filings dropdown, etc).

This could be investigated further, but IMO it's not the highest priority in this UI, not even in the top 3-4 (ie, behind Filing History, Todo List, App.vue, local filings, and more).

File: src/components/EntityInfo/EntityMenu.vue (hunk 20 above)
Comment (severinbeauvais) [reply src/components/EntityInfo/EntityMenu.vue:73]: https://app.zenhub.com/workspaces/entities-team-space-6143567664fb320019b81f39/issues/gh/bcgov/entity/15726

**Comment by severinbeauvais** [comment]
@jonathan-longe Thanks for all your forward-looking comments -- I really appreciate (and support) your efforts to move us towards a more flexible architecture. However, this ticket/PR was only meant to add a specific functionality using the existing architecture -- I am using ticket 15692 to refactor a variety of things, but even then the scope needs to be limited so we can balance cleanup vs new features.

I heard we have support to spend more time up front on this (pay now instead of pay later) but I'm trying to be careful to be concise in approved tickets, and creating other tickets for future work (or not creating tickets if the value isn't clear today -- they can be created later). For example, THIS ticket was estimated as a 3 and it's already used up 5 worth of time. And 15692 was also estimated as a 3 and maybe should be a 5 (due to revamped unit tests and QA). I'm OK with more tickets but I don't want to blow up current tickets.

Please continue your thorough reviews and spring of ideas, and please don't be upset if I push back a little :)

**Comment by severinbeauvais** [comment]
I'm not going to overrule you and merge this ticket as is. Please let me know if I have your approval (and if you're comfortable that outstanding issues are recorded in other tickets).

**Comment by jonathan-longe** [comment]
Understood @severinbeauvais I definitely don't want to hold you up or become a PITA :-)  Please go ahead and merge the code you're comfortable merging.
